dotnet/src/SemanticKernel.Core/Kernel.cs

```diff
@@ -107,70 +114,53 @@ public sealed class Kernel : IKernel, IDisposable
         this._logger = loggerFactory.CreateLogger(typeof(Kernel));
     }
 
-    /// <inheritdoc/>
-    [EditorBrowsable(EditorBrowsableState.Never)]
-    [Obsolete("This will be removed in a future release. Use RegisterPlugin instead.")]
-    public ISKFunction RegisterCustomFunction(ISKFunction customFunction)
+    /// <summary>
+    /// Kernel constructor. See KernelBuilder for an easier and less error prone approach to create kernel instances.
+    /// </summary>
+    /// <param name="aiServiceProvider">AI Service Provider</param>
+    /// <param name="plugins">The plugins.</param>
+    /// <param name="serviceSelector">AI Service selector</param>
+    /// <param name="httpHandlerFactory">HTTP handler factory</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    public Kernel(
+        IAIServiceProvider aiServiceProvider,
+        ISKPluginCollection? plugins = null,
+        IAIServiceSelector? serviceSelector = null,
+        IDelegatingHandlerFactory? httpHandlerFactory = null,
+        ILoggerFactory? loggerFactory = null)
     {
-        Verify.NotNull(customFunction);
-
-        this.Plugins.Add(new SKPlugin($"plugin{Guid.NewGuid():N}", new[] { customFunction }));
+        this.LoggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+        this.HttpHandlerFactory = httpHandlerFactory ?? NullHttpHandlerFactory.Instance;
+        this._aiServiceProvider = aiServiceProvider;
+        this.Plugins = plugins ?? new SKPluginCollection();
+        this._aiServiceSelector = serviceSelector ?? new OrderedIAIServiceSelector();
```

Comment:
We could make these cheaper by creating them lazily; that way they won't be instantiated unless used.  But if we expect not using them to be rare, it's fine and simpler to create them here in the ctor.

Reply:
A good idea that I would suggest deferring for a later stage after we have completed cloning.